### PR TITLE
Refactor pipeline runtime code

### PR DESCRIPTION
**Pull request by 6543**
next step in making code more structured and maintainable

- split pipeline exec into small functions
- fix edge-cases
- have less hidden code-flows
- fix possible [race condition](https://github.com/woodpecker-ci/woodpecker/pull/6272)

image

## Test Coverage Update: `pipeline/runtime`

**Overall: 80.5% → 98.3%** (+17.8pp)

| Function | `main` | `refactor` | Δ |
|---|--:|--:|--:|
| `GetShutdownCtx` | 0.0% | 100.0% | +100.0 |
| `WithDescription` | 0.0% | 100.0% | +100.0 |
| `makeLogger` | 75.0% | 100.0% | +25.0 |
| `Run` | 74.2% | 100.0% | +25.8 |
| `traceStep` | 84.6% | 100.0% | +15.4 |
| `completeStep` | — | 100.0% | ✱ new |
| `executeStep` | — | 100.0% | ✱ new |
| `logStages` | — | 100.0% | ✱ new |
| `runBlockingStep` | — | 100.0% | ✱ new |
| `runDetachedStep` | — | 85.7% | ✱ new |
| `runStage` | — | 100.0% | ✱ new |
| `shouldSkipStep` | — | 100.0% | ✱ new |
| `startStep` | — | 100.0% | ✱ new |
| `traceWorkflowSetupError` | — | 100.0% | ✱ new |
| `validateConfig` | — | 85.7% | ✱ new |

### CODE REVIEW

File: pipeline/runtime/step.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+		return nil
+	}
+
+	// Emit a "step started" trace before doing any real work.
+	if err := r.traceStep(nil, nil, step); err != nil {
+		return err
+	}
+
+	// Add compatibility environment variables for drone-ci plugins.
+	metadata.SetDroneEnviron(step.Environment)
```
Comment (6543) [inline pipeline/runtime/step.go]: we can do that only to plugins...

File: pipeline/runtime/step.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+// executeStep is the single entry point called per step from runStage.
+// It checks whether the step should be skipped, emits a "started" trace,
+// sets up drone-compat env vars, then hands off to blocking or detached execution.
+func (r *Runtime) executeStep(runnerCtx context.Context, step *backend.Step) error {
+	logger := r.makeLogger()
+	logger.Debug().Str("step", step.Name).Msg("prepare")
+
+	if r.shouldSkipStep(step) {
+		return nil
```
Comment (6543) [inline pipeline/runtime/step.go]: should we not update the step state to skiped and report via traceStep?

File: pipeline/runtime/workflow.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+// Run starts the workflow, executes all stages sequentially, and tears down the
+// workflow on exit. The runnerCtx must outlive workflow cancellation so that cleanup
+// can still reach the backend (e.g. stopping Docker containers).
+func (r *Runtime) Run(runnerCtx context.Context) error {
+	logger := r.makeLogger()
+	r.logStages()
+
+	defer func() {
```
Comment (6543) [inline pipeline/runtime/workflow.go:43]:
```suggestion
	// we make sure cleanup always happens
	defer func() {
```

File: agent/tracer.go [hunk 4]
```diff
@@ -48,7 +48,7 @@ func (r *Runner) createTracer(ctxMeta context.Context, uploads *sync.WaitGroup,
 			Exited:   state.Process.Exited,
 			ExitCode: state.Process.ExitCode,
 			Started:  state.Process.Started,
-			Canceled: errors.Is(state.Process.Error, pipeline_errors.ErrCancel),
+			Canceled: errors.Is(state.Process.Error, pipeline_errors.ErrCancel) || state.Process.Skipped,
```
Comment (6543) [inline agent/tracer.go]:
```suggestion
			Canceled: errors.Is(state.Process.Error, pipeline_errors.ErrCancel) || state.Skipped,
```

**Comment by 6543** [comment]
```
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:29:	WithBackend		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:36:	WithLogger		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:43:	WithTracer		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:50:	WithContext		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:57:	WithDescription		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:64:	WithTaskUUID		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/runtime.go:53:	New			100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/runtime.go:67:	makeLogger		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/shutdown.go:33:	GetShutdownCtx		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:32:		executeStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:63:		shouldSkipStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:95:		startStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:126:	completeStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:174:	runBlockingStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:202:	runDetachedStep		85.7%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:239:	traceStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/workflow.go:34:	Run			100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/workflow.go:80:	logStages		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/workflow.go:98:	traceWorkflowSetupError	100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/workflow.go:121:	runStage		100.0%
total:									(statements)		98.8%
```

# -> 98.8% coverage :tada:

**Comment by 6543** [comment]
```
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:29:	WithBackend		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:36:	WithLogger		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:43:	WithTracer		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:50:	WithContext		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:57:	WithDescription		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/option.go:64:	WithTaskUUID		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/runtime.go:59:	New			100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/runtime.go:73:	makeLogger		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/shutdown.go:33:	GetShutdownCtx		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:32:		executeStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:63:		shouldSkipStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:95:		startStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:126:	completeStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:174:	runBlockingStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:203:	runDetachedStep		94.1%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/step.go:247:	traceStep		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/workflow.go:34:	Run			100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/workflow.go:85:	logStages		100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/workflow.go:103:	traceWorkflowSetupError	100.0%
go.woodpecker-ci.org/woodpecker/v3/pipeline/runtime/workflow.go:126:	runStage		100.0%
total:									(statements)		99.4%
```

# -> 99.4% coverage 🎉

and some fixes ;)

**Comment by 6543** [comment]
the last commit got me more test coverage ... but now services dont get cancled anymore and wait indefinetly ...

so i move that into it's own pull as it is significat enouth to do so ... -> https://github.com/woodpecker-ci/woodpecker/pull/6263

and revert it here

File: pipeline/state/state.go [hunk 5]
```diff
@@ -18,20 +18,21 @@ import (
...
-		Process backend.State
+// State is used to signal the current workflow and step state.
+// Only steps using the trace func report back what's going on.
+// And the workflow is updated alongside it.
+type State struct {
+	// Global state of the currently running Workflow.
+	Workflow struct {
+		// Workflow start time
+		Started int64 `json:"time"`
+		// Current pipeline error state
```
Comment (anbraten) [inline pipeline/state/state.go]:
```suggestion
		// Current workflow error state
```

File: pipeline/state/state.go [hunk 6]
```diff
@@ -18,20 +18,21 @@ import (
...
+		Error error `json:"error"`
 	}
-)
+
+	// Current step that updates the step and workflow state
+	CurrStep *backend.Step `json:"step"`
+
+	// Current Step state.
```
Comment (anbraten) [inline pipeline/state/state.go]:
```suggestion
	// Current step state
```

File: pipeline/runtime/workflow_test.go [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+
+func TestMakeLoggerWithDescription(t *testing.T) {
+	t.Parallel()
+	r := New(&backend.Config{},
+		WithBackend(dummy.New()),
+		WithTracer(newTestTracer(t)),
+		WithLogger(newTestLogger(t)),
+		WithDescription(map[string]string{"repo": "woodpecker", "branch": "main"}),
+	)
+	r.logStages()
```
Comment (anbraten) [inline pipeline/runtime/workflow_test.go]: Anything we can test here

File: agent/tracer.go [hunk 8]
```diff
@@ -68,21 +68,21 @@ func (r *Runner) createTracer(ctxMeta context.Context, uploads *sync.WaitGroup,
...
 
 		// TODO: find better way to update this state and move it to pipeline to have the same env in cli-exec
-		state.Pipeline.Step.Environment["CI_MACHINE"] = r.hostname
+		state.CurrStep.Environment["CI_MACHINE"] = r.hostname
 
-		state.Pipeline.Step.Environment["CI_PIPELINE_STARTED"] = strconv.FormatInt(state.Pipeline.Started, 10)
+		state.CurrStep.Environment["CI_PIPELINE_STARTED"] = strconv.FormatInt(state.Workflow.Started, 10)
 
-		state.Pipeline.Step.Environment["CI_STEP_STARTED"] = strconv.FormatInt(state.Pipeline.Started, 10)
+		state.CurrStep.Environment["CI_STEP_STARTED"] = strconv.FormatInt(state.Workflow.Started, 10)
```
[the remark this answers is not in the document]
Comment (6543) [reply agent/tracer.go:84]: known bug not part of this pull scope

File: pipeline/runtime/step.go [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+// If StartStep or TailStep fail, startStep returns a non-nil error and the caller
+// must not call waitForLogs.
+func (r *Runtime) startStep(step *backend.Step) (func(), int64, error) {
+	if err := r.engine.StartStep(r.ctx, step, r.taskUUID); err != nil {
+		return nil, 0, err
+	}
+	startTime := time.Now().Unix()
+
+	rc, err := r.engine.TailStep(r.ctx, step, r.taskUUID)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (6543) [reply pipeline/runtime/step.go:102]: no it wont... DestroyWorkflow will as per interface contract clean up all things left

File: pipeline/runtime/step.go [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+	s.Workflow.Error = r.err.Get()
+
+	switch {
+	case processState == nil && err != nil:
+		// Step failed to start — synthesis an exited process state.
+		s.CurrStepState = backend.State{
+			Error:     err,
+			Exited:    true,
+			OOMKilled: false,
+		}
```
[the remark this answers is not in the document]
Comment (6543) [reply pipeline/runtime/step.go:250]: i have no idea why this word is so liked by AI ... propose a new sentence without it!

File: pipeline/runtime/workflow.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+// can still reach the backend (e.g. stopping Docker containers).
+func (r *Runtime) Run(runnerCtx context.Context) error {
+	logger := r.makeLogger()
+
+	if r.tracer == nil {
+		return fmt.Errorf("runtime misconfiguration: tracer must not be nil")
+	}
+	if r.logger == nil {
+		return fmt.Errorf("runtime misconfiguration: logger must not be nil")
+	}
```
[the remark this answers is not in the document]
Comment (6543) [reply pipeline/runtime/workflow.go:37]: made a proper refactor: a7780e3b0c1c83457f951528967c4bd34df0368f

File: pipeline/runtime/workflow.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+	for _, stage := range r.spec.Stages {
+		select {
+		case <-r.ctx.Done():
+			return pipeline_errors.ErrCancel
+		case err := <-r.runStage(runnerCtx, stage.Steps):
+			if err != nil {
+				r.err.Set(err)
+			}
```
[the remark this answers is not in the document]
Comment (6543) [reply pipeline/runtime/workflow.go:68]: its fine as is

File: pipeline/tracing/tracer.go [hunk 13]
```diff
@@ -38,15 +38,15 @@ func (f TraceFunc) Trace(state *state.State) error {
...
 	}
-	if state.Pipeline.Step.Environment == nil {
+	if state.CurrStep.Environment == nil {
 		return nil
 	}
-	state.Pipeline.Step.Environment["CI_PIPELINE_STARTED"] = strconv.FormatInt(state.Pipeline.Started, 10)
+	state.CurrStep.Environment["CI_PIPELINE_STARTED"] = strconv.FormatInt(state.Workflow.Started, 10)
 
-	state.Pipeline.Step.Environment["CI_STEP_STARTED"] = strconv.FormatInt(state.Pipeline.Started, 10)
+	state.CurrStep.Environment["CI_STEP_STARTED"] = strconv.FormatInt(state.Workflow.Started, 10)
```
[the remark this answers is not in the document]
Comment (6543) [reply pipeline/tracing/tracer.go:49]: didnt i tell you that you should ignore it?!?

File: pipeline/runtime/workflow.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+// traceWorkflowSetupError traces an ErrInvalidWorkflowSetup to the tracer, if one
+// is configured. Other error types are silently ignored here (they are still
+// returned by Run).
+func (r *Runtime) traceWorkflowSetupError(err error) {
+	var stepErr *pipeline_errors.ErrInvalidWorkflowSetup
+	if !errors.As(err, &stepErr) {
+		return
+	}
+
```
[the remark this answers is not in the document]
Comment (6543) [reply pipeline/runtime/workflow.go]: done

File: server/pipeline/step_status_test.go [hunk 15]
```diff
@@ -216,6 +216,29 @@ func TestUpdateStepStatus(t *testing.T) {
...
+			t.Parallel()
+			step := &model.Step{State: model.StatusPending}
+			state := rpc.StepState{Canceled: true, Exited: true, Finished: 50}
+
+			err := UpdateStepStatus(t.Context(), mockStoreStep(t), step, state)
+
+			assert.NoError(t, err)
+			assert.Equal(t, model.StatusKilled, step.State)
+			assert.Equal(t, int64(50), step.Finished)
+		})
```
[the remark this answers is not in the document]
Comment (6543) [reply server/pipeline/step_status_test.go:271]: let it be for now

File: pipeline/runtime/workflow_test.go (hunk 7 above)
Comment (6543) [reply pipeline/runtime/workflow_test.go]: if it panics

Comment (6543) [reply pipeline/runtime/workflow_test.go]: the tests where mostly written by claude, through i did a ton of work to make it fit and work and follow our guidelines etc ... and post processin it myselfe afterwards ...

anyway, this one the ai just mostly added for code coverage ... but i let it stay as it still could panic in nil ref errors ...

**Comment by 6543** [comment]
looks like paralel test is still flaky :/

https://ci.woodpecker-ci.org/repos/3780/pipeline/32422/37

File: server/pipeline/step_status.go [hunk 16]
```diff
@@ -34,8 +34,10 @@ func UpdateStepStatus(ctx context.Context, store store.Store, step *model.Step,
 
 	switch step.State {
 	case model.StatusPending:
-		// Transition from pending to running when started
-		if state.Finished == 0 {
+		// Transition from pending to running or skipped
+		if state.Canceled {
+			step.State = model.StatusSkipped
```
Comment (6543) [inline server/pipeline/step_status.go:40]: cancled has its own state now meaning now -> https://github.com/woodpecker-ci/woodpecker/pull/6176 we need to acount for that

**Comment by 6543** [comment]
moved the integration test too #6270 ...

**Comment by 6543** [comment]
I'll move skip forward out too in an own pull: https://github.com/woodpecker-ci/woodpecker/pull/6295

File: server/pipeline/step_status.go (hunk 16 above)
Comment (6543) [reply server/pipeline/step_status.go:40]: fixed in #6295

**Comment by 6543** [comment]
ok now only the related refactoring remains :)

File: pipeline/runtime/workflow_test.go (hunk 7 above)
Comment (6543) [reply pipeline/runtime/workflow_test.go]: @anbraten if you want i can remove it...

**Comment by 6543** [comment]
not sure if we wana save the flowchart somewhere?
[woodpecker_workflow_run_flowchart.svg](https://github.com/user-attachments/assets/2000ec47-5e06-4073-93fa-ec38c0ab1c47)

File: pipeline/runtime/step.go [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+		logger := r.makeLogger()
+
+		processState, err := r.completeStep(runnerCtx, step, waitForLogs, startTime)
+		logger.Debug().Str("step", step.Name).Msg("complete")
+
+		if errors.Is(err, context.Canceled) {
+			err = pipeline_errors.ErrCancel
+		}
+		if err != nil {
+			logger.Error().Err(err).Str("step", step.Name).Msg("detached step failed after setup")
```
Comment (6543) [inline pipeline/runtime/step.go]:
```suggestion
			logger.Error().Err(err).Str("step", step.Name).Msg("detached step failed after while running")
```

**Comment by qwerty287** [comment]
> not sure if we wana save the flowchart somewhere? [woodpecker_workflow_run_flowchart.svg](https://github.com/user-attachments/assets/2000ec47-5e06-4073-93fa-ec38c0ab1c47)

Doesn't hurt I think, maybe here? https://woodpecker-ci.org/docs/development/architecture#agent

**Comment by 6543** [comment]
@qwerty287 done

File: pipeline/runtime/workflow_test.go (hunk 7 above)
Comment (6543) [reply pipeline/runtime/workflow_test.go]: ok i just removed it for now... d76dc64cb8ef32838688ee060aef74d22278c370

**Comment by 6543** [comment]
@anbraten all reviews addressed

File: pipeline/runtime/step.go [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+
+	rc, err := r.engine.TailStep(r.ctx, step, r.taskUUID)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
```
Comment (lafriks) [inline pipeline/runtime/step.go]: Maybe replace this with Go 1.25 Go function:

```go
wg.Go(func() {
```

Comment (6543) [reply pipeline/runtime/step.go]: uh did not notice they added that :)

will look into the codebase :rocket:

File: pipeline/runtime/workflow.go [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// runStage executes all steps of a stage in parallel.
+// It returns a channel that emits the combined error (if any) once all steps finish.
+func (r *Runtime) runStage(runnerCtx context.Context, steps []*backend_types.Step) <-chan error {
+	var g errgroup.Group
+	done := make(chan error)
+
+	for _, step := range steps {
+		step := step // capture loop variable
```
Comment (lafriks) [inline pipeline/runtime/workflow.go]: Isn't that done already starting from 1.24?

Comment (6543) [reply pipeline/runtime/workflow.go]: well the old had it, the new had it too...
https://github.com/woodpecker-ci/woodpecker/blob/4b4ef5e591e7f0281e809aab9e7b4d9cf7a7e927/pipeline/runtime/executor.go#L137

but yes we can refactor that too

Comment (6543) [reply pipeline/runtime/workflow.go]: 3739bcd33dc89a978f3aa6e804a6a6939bf16fd3
